rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java

```diff
@@ -506,172 +510,150 @@ public void onFailure(Throwable caught) {
+                                                          " for input: " + jsNodes.get(i).toString());
             }
         }
 
-        model.setNumBusy(numBusy);
-        model.setNumConfiguring(numConfiguring);
-        model.setNumDeploying(numDeploying);
-        model.setNumDown(numDown);
-        model.setNumFree(numFree);
-        model.setNumLost(numLost);
-        model.setNumToBeRemoved(numToBeRemoved);
+        model.setNodes(newNodeSources);
 
-        model.setNumLocked(numLocked);
+        recalculatePhysicalVirtualHosts();
+
+        recalculateStatistics();
+
+    }
+
+    /**
+     * Decides whether it was regular server request, or not.
+     * 'Not regular' request means that something went strange, e.g. server was restarted,
+     * or it is the very first request of this type.
+     * It it is regular, than rm portal should treat server response as a delta.
+     * If it is not regular, than rm portal should 'forget' what it knew, and start from the scratch
+     * @param currentCounter is current counter that client is aware of
+     * @param latestCounter is counter received from server
+     * @return true if request is regular
+     */
+    private boolean isRegularRequest(long currentCounter, Long latestCounter) {
+        return 0 <= currentCounter && currentCounter <= latestCounter;
+    }
+
+    /**
+     * Add and remove NodeSources
+     * @param newNodeSources
+     * @param obj
+     */
+    private void processNodeSources(HashMap<String, NodeSource> newNodeSources, JSONObject obj) {
+        JSONArray jsNodeSources = obj.get("nodeSource").isArray();
+        for (int i = 0; i < jsNodeSources.size(); i++) {
+            JSONObject jsNodeSource = jsNodeSources.get(i).isObject();
+
+            NodeSource nodeSource = parseNodeSource(jsNodeSource);
+            if (nodeSource.isRemoved()) {
+                newNodeSources.remove(nodeSource.getSourceName());
+            } else {
+                newNodeSources.put(nodeSource.getSourceName(), nodeSource);
+            }
+        }
+    }
+
+    private void removeNodeFromNodeSource(Node node, NodeSource nodeSource) {
+        if (node.isDeployingNode()) {
+            nodeSource.getDeploying().remove(node.getNodeUrl());
+        } else {
+            Host host = nodeSource.getHosts().get(node.getHostName());
+
+            if (host != null) {
+                host.getNodes().remove(node.getNodeUrl());
+
+                if (host.getNodes().isEmpty()) {
+                    nodeSource.getHosts().remove(host);
+                }
+            }
+        }
+    }
 
+    private void addNodeToNodeSource(Node node, NodeSource nodeSource) {
+        // as deploying node
+        if (node.isDeployingNode()) {
+
+            nodeSource.getDeploying().put(node.getNodeUrl(), node);
+
+        } else { // as already deployed node
+            Host host = nodeSource.getHosts().get(node.getHostName());
+
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Add a default case to this switch. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)